### PR TITLE
Add support for folding SignOp in --stablehlo-refine-shapes

### PR DESCRIPTION
**Pull request by burmako**
This follows up on the recent DynamicConvOp pull request and introduces additional functionality to handle accompanying shape computations.

### CODE REVIEW

**Comment by burmako** [comment]
MLIR-HLO commit: https://github.com/tensorflow/mlir-hlo/commit/3007ba96161759a1a45a1f17d04d141ba472e676.

File: stablehlo/transforms/StablehloRefineShapes.cpp [hunk 1]
```diff
@@ -314,6 +327,25 @@ struct EvalSelectOpPattern : public OpRewritePattern<SelectOp> {
...
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(SignOp op,
+                                PatternRewriter& rewriter) const override {
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+    auto resultBitwidth = resultType.getElementType().getIntOrFloatBitWidth();
+    return evalUnary(rewriter, op, [&](APInt operand) {
+      int64_t result;
+      if (operand.slt(0))
+        result = -1;
+      else if (operand.getSExtValue() == 0)
```
Comment (GleasonK) [inline stablehlo/transforms/StablehloRefineShapes.cpp:340]: Could consider `APInt::isNegative`, `APInt::isStrictlyPositive`, `APInt::isZero`

File: stablehlo/transforms/StablehloRefineShapes.cpp [hunk 2]
```diff
@@ -191,27 +218,13 @@ struct EvalConvertOpPattern : public OpRewritePattern<ConvertOp> {
...
-      result.push_back(resultEl);
-    }
-
-    rewriter.replaceOpWithNewOp<ConstantOp>(
-        op, DenseIntElementsAttr::get(resultType, result));
-    return success();
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+    auto resultBitwidth = resultType.getElementType().getIntOrFloatBitWidth();
+    return evalUnary(rewriter, op, [&](APInt operand) {
+      return APSInt(operand, isOperandUnsigned).extOrTrunc(resultBitwidth);
```
Comment (GleasonK) [inline stablehlo/transforms/StablehloRefineShapes.cpp:226]: Lets make sure this matches the reference impl once we have one. Quantization may use ConvertOp sometimes, would just want to make sure it matches up.

**Review by GleasonK — APPROVED** [review]

File: stablehlo/transforms/StablehloRefineShapes.cpp (hunk 1 above)
Comment (burmako) [reply stablehlo/transforms/StablehloRefineShapes.cpp]: Nice!

File: stablehlo/transforms/StablehloRefineShapes.cpp (hunk 2 above)
Comment (burmako) [reply stablehlo/transforms/StablehloRefineShapes.cpp:226]: That's a good point! I've made a note.

**Pull request merged**